core/src/avm2/array.rs

```diff
@@ -33,49 +224,189 @@ impl<'gc> ArrayStorage<'gc> {
+            }
+            ArrayStorage::Sparse(..) => {}
+        }
+    }
+
+    fn convert_to_dense(&mut self) {
+        match self {
+            ArrayStorage::Dense(..) => {}
+            ArrayStorage::Sparse(storage, length) => {
+                let mut new_storage = Vec::new();
+                for i in 0..*length {
+                    let value = storage.get(&i).copied();
+                    new_storage.push(value);
+                }
+                let dense_used = new_storage.iter().filter(|v| v.is_some()).count();
+                *self = ArrayStorage::Dense(new_storage, dense_used);
+            }
+        }
     }
 
     /// Set an array storage slot to a particular value.
     ///
     /// If the item index extends beyond the length of the array, then the
     /// array will be extended with holes.
     pub fn set(&mut self, item: usize, value: Value<'gc>) {
-        if self.storage.len() < (item + 1) {
-            self.storage.resize(item + 1, None)
+        match self {
+            ArrayStorage::Dense(storage, dense_used) => {
+                if storage.len() < (item + 1) {
+                    if *dense_used < (item / 4) && MIN_SPARSE_LENGTH < item
+                        || item >= MAX_DENSE_LENGTH
+                    {
+                        self.convert_to_sparse();
+                        if let ArrayStorage::Sparse(storage, length) = self {
+                            *length = item + 1;
+                            storage.insert(item, value);
+                        }
+                    } else {
+                        storage.resize(item + 1, None);
+                        *storage.get_mut(item).unwrap() = Some(value);
+                        let new_holes = (item + 1) - storage.len();
+                        *dense_used = (*dense_used + new_holes) + 1;
+                    }
+                } else {
+                    if storage[item].is_none() {
+                        *dense_used += 1;
+                    }
+                    *storage.get_mut(item).unwrap() = Some(value);
+                }
+            }
+            ArrayStorage::Sparse(storage, length) => {
+                storage.insert(item, value);
+                if item >= *length {
+                    *length = item + 1;
+                }
+            }
         }
-
-        *self.storage.get_mut(item).unwrap() = Some(value)
     }
 
     /// Delete an array storage slot, leaving a hole.
     pub fn delete(&mut self, item: usize) {
-        if let Some(i) = self.storage.get_mut(item) {
-            *i = None;
+        match self {
+            ArrayStorage::Dense(storage, dense_used) => {
+                if let Some(i) = storage.get_mut(item) {
+                    *i = None;
+                    if *dense_used != 0 {
+                        *dense_used -= 1;
+                        if *dense_used == 0 {
+                            self.convert_to_dense();
```

Comment:
How is this supposed to work? The array storage is already `Dense` here.

Reply:
Oops, I don't know how I overlooked that, it's not supposed to be there. Now I've added conversion to dense if the sparse array is empty.